#102342692 cancel reservation

elchroy opened the pull request:
#### What does this PR do?

This PR enables a resly user to cancel a reservation.
#### Description of Task to be completed?

The user can cancel a reservation which he/her has already paid for.
This makes a new refund tied to that user and to that particular reservation.
The refund is a percentage of the total cost of that reservation.
#### How should this be manually tested?

After login and creating a reservation,

> Landing page - Your profile.
> Make a reservation and visit the restaurant.
> Current Reservations.
> Hit the `x` button to cancel a reservation.
> Confirm that the reservation is to be cancelled.
> Goto Cancelled reservations page and view the cancelled reservations.
#### Any background context you want to provide?

A user that is not able to make it for his reservation might wish to cancel it. This is now possible but the user will only be refunded a percentage of the cost of the reservation.
A user cannot cancel a reservation that is not tied to himself/herself.
A user cannot cancel a reservation that is due in the next 15 minutes.
A user cannot cancel a reservation that is passed.
#### What are the relevant pivotal tracker stories?

The Pivotal Tracker for this story can be viewed [here](https://www.pivotaltracker.com/projects/1415814/stories/102342692).
#### Screenshots (if appropriate)

![screen shot 2016-04-11 at 11 44 23 pm](https://cloud.githubusercontent.com/assets/17028608/14462841/bf96b7da-00c0-11e6-866b-de7ff9a31ab4.png)
![screen shot 2016-04-11 at 11 44 38 pm](https://cloud.githubusercontent.com/assets/17028608/14462846/bf989c1c-00c0-11e6-96d3-f840924502b3.png)
![screen shot 2016-04-11 at 11 45 07 pm](https://cloud.githubusercontent.com/assets/17028608/14462842/bf97083e-00c0-11e6-8ef6-7ea204c44af1.png)
![screen shot 2016-04-11 at 11 45 34 pm](https://cloud.githubusercontent.com/assets/17028608/14462845/bf97b9e6-00c0-11e6-80c0-fe820bc2a8fd.png)
![screen shot 2016-04-11 at 11 46 17 pm](https://cloud.githubusercontent.com/assets/17028608/14462843/bf97664e-00c0-11e6-9bdb-ce4f9a196169.png)
![screen shot 2016-04-11 at 11 46 45 pm](https://cloud.githubusercontent.com/assets/17028608/14462844/bf97afbe-00c0-11e6-8c6a-61619fb25826.png)
![screen shot 2016-04-12 at 3 13 21 pm](https://cloud.githubusercontent.com/assets/17028608/14462989/4ad1af6c-00c1-11e6-93da-b6bd05b2a529.png)
#### Questions:

There are no questions at this time.

elchroy committed:
[Feature #102342692] Cancel Reservation

A diner should be able to receive a partial refund on a cancelled reservation.
The diner is given only a percentage of the amount paid.
The refund is stored as credits to the user.

Finishes #102342692

elchroy committed:
[Feature #102342692] Cancel Reservation

A diner can cancel a reservation and will get a partial refund.
The refund is aded for the user in the Refund model.
The reservations are also updated on the cancelled reservations page.

Finishes #102342692

elchroy committed:
[HotFix #102342692] Cancel Reservation

The return message for the successful cancellation of a restaurant is corrected.

Finishes #102342692

andela-anandaa commented on app/Http/Controllers/BookingController.php:
```diff
@@ -265,6 +266,54 @@ public function delteCartItem($item_id)
+        $res = Booking::find($id);
+        if ($res == null) {
+            $output['status'] = 'failure';
+            $output['message'] = 'Booking not found.';
```
Proper way of declaring this array:

```
$output = [
    'status' => 'failure',
    'messages' => 'Booking not found.',
];
```

andela-anandaa commented on app/Http/Controllers/BookingController.php:
```diff
@@ -265,6 +266,54 @@ public function delteCartItem($item_id)
+
+    private function getRefund($cost)
+    {
+        return $cost * 0.7;
```
Will be better to have this kind of settings in a table in the DB.

andela-anandaa commented on app/Http/Controllers/BookingController.php:
```diff
@@ -265,6 +266,54 @@ public function delteCartItem($item_id)
+        }
+        $timeOff = $request->input('offset');
+        if ($res->isSoon($timeOff)) {
+            $output['status'] = 'failure';
```
Same here.

andela-anandaa commented on app/Http/Controllers/BookingController.php:
```diff
@@ -265,6 +266,54 @@ public function delteCartItem($item_id)
+            $output['message'] = 'This reservation has passed.';
+            return json_encode($output);
+        }
+        $refund = new Refund;
```
`new Refund()`

andela-anandaa commented on public/js/refund.js:
```diff
@@ -0,0 +1,61 @@
+        for (var i=0; i<user_rating; i++) {
+            user_rating_output += "<i class='fa fa-star' style='color:#FFD700'></i>";
+        }
+        $('#user_rating_'+res+'_'+booking).html(user_rating_output);
```
There should be a space before and after `+`.

andela-anandaa commented on public/js/refund.js:
```diff
@@ -0,0 +1,61 @@
+            dialogClass: "modal-dialog modal-lg" // Bootstrap classes for large modal
+    });
+
+    var updateStars = function (id) {
```
Remove.

elchroy committed:
[Chores #102342692] Cancel Reservation

The return output array is modified.
The cancel link is changed to a cancel button.
After checking out the user is redirected to his current reservations.
String concatenation now includes spaces to follow standard.

Finishes #102342692

andela-anandaa commented:
@andela-celisha-wigwe - Fix the failing tests.

andela-anandaa commented:
@andela-celisha-wigwe - Fix the merge-conflict then I will merge.

andela-anandaa commented on app/Booking.php:
```diff
@@ -25,6 +26,31 @@ public function user()
+        return $this->hasOne('Resly\Refund');
+    }
+
+    public function isSoon($timeOff)
```
Write a comment here too, on what really "soon" means.

andela-anandaa replied:
Write a comment here too, on what really "soon" means.

andela-anandaa commented on app/Http/Controllers/BookingController.php:
```diff
@@ -188,15 +189,148 @@ public function book($restaurant_id)
+        });
+        $count = count($bookedTablesIDs);
+
+        $bookedTables = array_map(function ($bookedTable) { return Table::find($bookedTable);}, $bookedTablesIDs);
```
It's better readable written this way:

``` php
array_map(function ($bookedTable) { 
    return Table::find($bookedTable);
}, $bookedTablesIDs);
```

``

andela-anandaa commented on app/Http/Controllers/BookingController.php:
```diff
@@ -188,15 +189,148 @@ public function book($restaurant_id)
+
         return view('bookings.book')
-        ->with('restaurant', $restaurant);
+        ->with(['restaurant' => $restaurant, 'bookedTables' => $bookedTables, 'tableIDs' => $bookedTablesIDs, 'count' => $count]);
```
``` php
[
    'restaurant' => $restaurant, 
    'bookedTables' => $bookedTables, 
    'tableIDs' => $bookedTablesIDs, 
    'count' => $count,
];
```

andela-anandaa commented on app/Http/Controllers/BookingController.php:
```diff
@@ -188,15 +189,148 @@ public function book($restaurant_id)
+    {
+        $user = Auth::user()->id . Auth::user()->username;
+        $tables = $request->session()->get($user, function () {
+            return array();
```
Or `[]`

andela-anandaa commented on app/Http/Controllers/BookingController.php:
```diff
@@ -188,15 +189,148 @@ public function book($restaurant_id)
+        $interval = $allowedBookingDateTime->diff($bookingDateTime);
+
+        // Calculate the seconds difference between the allowedBookingDateTime and bookingDateTime
+        // Reference: http://stackoverflow.com/questions/14277611/convert-dateinterval-object-to-seconds-in-php
```
Good context for your approach, esp. with S/O reference.

andela-anandaa commented on app/Http/Controllers/BookingController.php:
```diff
@@ -188,15 +189,148 @@ public function book($restaurant_id)
+
+
+        if ($seconds < 0) {
+            // $output['message'] = "Specify a date at least 30 minutes later.";
```
Always remove any commented out code before PR.

andela-anandaa commented on app/Http/Controllers/BookingController.php:
```diff
@@ -188,15 +189,148 @@ public function book($restaurant_id)
     public function addTable(Request $request)
     {
+        // return json_encode("hihihhih");
+
```
I don't understand this code being here -- this should be handled in the authentication middleware layer at the _routes_.

andela-anandaa commented on app/Http/Controllers/BookingController.php:
```diff
@@ -232,25 +368,72 @@ public function addTable(Request $request)
+        // return redirect()->back();
+    }
+
+    private function prepareCartData($request ,$table)
```
What was the essence of separating out this method? Looks like you had some ideas.

andela-anandaa commented on app/Http/Controllers/TableController.php:
```diff
@@ -60,6 +61,34 @@ public function edit(Request $request)
+        });
+        $ids = array_values($tables);
+        $bookedTables = array_map(function ($id) { return $this->tableRepository->get($id);}, $ids);
+        array_map(function ($bookedTable) {$bookedTable->cost = number_format($bookedTable->cost, 2);}, $bookedTables);
```
Do the same, as recommended previously. Always avoid stacking things on one line even, esp. with functions that all call-backs.

andela-anandaa commented on app/Http/Controllers/TableController.php:
```diff
@@ -60,6 +61,34 @@ public function edit(Request $request)
+        array_map(function ($bookedTable) {$bookedTable->cost = number_format($bookedTable->cost, 2);}, $bookedTables);
+        $sum = 0;
+        $total_seats = 0;
+        array_map(function ($bookedTable) use (&$sum, &$total_seats) { $sum = $sum + $bookedTable->cost; $total_seats = $total_seats + $bookedTable->seats_number;}, $bookedTables);
```
Same here.

andela-anandaa commented on public/css/book.css:
```diff
@@ -0,0 +1,40 @@
+        pointer : cursor;
+      }
+
+      .table-booked {
```
Clear the extra indentation on these lines, till line 40.

andela-anandaa commented on public/js/add_tables.js:
```diff
@@ -77,7 +77,7 @@ $(document).ready(function() {
     tableTrack.restaurant_id = parseInt($("#rest_id").val());
 
     console.dir(tableTrack);
```
This should be removed for production -- no consoling.

andela-anandaa commented on public/js/booking_modes.js:
```diff
@@ -0,0 +1,150 @@
+// JS to handle different modes of booking a table.
```
Is calling this file `boking_models.js` is the right? Does model have anything to do with MVC?

andela-anandaa commented:
Finally, write tests for your feature.

elchroy committed:
[Chores #102342692] Cancel Reservation

The user is redirected to the cancelled reservation page once a reservation has been cancelled.
The isSoon method is better described with a comment. This makes for better understanding.

elchroy committed:
[Fix #102342692] Cancel Reservation

The merge conflicts with the staging branch are fixed.

elchroy committed:
[Fix #102342692] Fix merge conflicts

Fix merge conflicts with the staging branch

elchroy committed:
[Chores #102342692] Add tests for cancel feature

The feature that enables users to cancel a reservation is tested
to ensure that the expected results are met.
